cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -0,0 +1,299 @@
+/*
+ * Copyright (c) 2020, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cmath>
+#include <cassert>
+#include <functional>
+#include <iostream>
+#include <limits>
+
+namespace cudf {
+namespace fixed_point {
+
+template <typename T>
+struct weak_typedef {
+    T _t;
+    explicit weak_typedef(T t) : _t(t) {}
+    operator T() const { return _t; }
+};
+
+using scale_type = weak_typedef<int32_t>;
+
+enum class Radix : int32_t {
+    BASE_2  = 2,
+    BASE_10 = 10
+};
+
+namespace detail {
+    // helper function to negate strongly typed scale_type
+    auto negate(scale_type const& scale) -> scale_type {
+        return scale_type{-scale};
+    }
+
+    // perform this operation when constructing with positive scale
+    template <Radix Rad, typename T>
+    constexpr auto right_shift(T const& val, scale_type const& scale) -> double {
+        assert(scale > 0);
+        return val / std::pow(static_cast<int32_t>(Rad), static_cast<int32_t>(scale));
+    }
+
+    // perform this operation when constructing with negative scale
+    template <Radix Rad, typename T>
+    constexpr auto left_shift(T const& val, scale_type const& scale) -> double {
+        assert(scale < 0);
+        return val * std::pow(static_cast<int32_t>(Rad), static_cast<int32_t>(negate(scale)));
+    }
+
+    // convenience generic shift function
+    template <Radix Rad, typename T>
+    constexpr auto shift(T const& val, scale_type const& scale) -> double {
+        if      (scale == 0) return static_cast<double>(val);
+        else if (scale >  0) return right_shift<Rad>(val, scale);
+        else                 return left_shift <Rad>(val, scale);
+    }
+}
+
+template <typename T>
+constexpr inline auto is_supported_representation_type() -> bool {
+  return std::is_same<T, int32_t>::value;
+}
+
+template <typename T>
+constexpr inline auto is_supported_construction_value_type() -> bool {
+  return std::is_integral      <T>::value
+      || std::is_floating_point<T>::value;
+}
+
+// helper struct for constructing fixed_point when value is already shifted
```

Comment:
I think "value is already shifted" is a bit unclear. Suggest providing an example. I guess it means "`value` is specified in the given scaling factor", as opposed to "value is unscaled".

Reply:
Done